src/Chemistry.hpp

```diff
@@ -0,0 +1,141 @@
+#ifndef CHEMISTRY_HPP_ // NOLINT
+#define CHEMISTRY_HPP_
+//==============================================================================
+// TwoMomentRad - a radiation transport library for patch-based AMR codes
+// Copyright 2020 Benjamin Wibking.
+// Released under the MIT license. See LICENSE file included in the GitHub repo.
+//==============================================================================
+/// \file CloudyCooling.hpp
+/// \brief Defines methods for interpolating cooling rates from Cloudy tables.
+///
+
+#include <limits>
+
+#include "AMReX.H"
+#include "AMReX_BLassert.H"
+#include "AMReX_Extension.H"
+#include "AMReX_GpuQualifiers.H"
+
+#include "hydro_system.hpp"
+#include "radiation_system.hpp"
+
+#include "burn_type.H"
+#include "burner.H"
+#include "eos.H"
+#include "extern_parameters.H"
+
+namespace quokka::chemistry
+{
+template <typename problem_t> void computeChemistry(amrex::MultiFab &mf, const Real dt_in)
+{
+	BL_PROFILE("computeChemistry()")
+
+	const Real grav_constant = 6.674e-8;
+	const Real dt = dt_in;
+
+	const auto &ba = mf.boxArray();
+	const auto &dmap = mf.DistributionMap();
+	amrex::iMultiFab nsubstepsMF(ba, dmap, 1, 0);
```

Comment:
This is never actually used within the `ParallelFor` loop. Should either get rid of it or set it appropriately.

Reply:
Do I just delete these lines? I don't know what you mean by 'set it appropriately'.

Reply:
In the cooling solve, this array is meant to store the number of substeps it takes to integrate each cell. I don't know how to get this quantity from Microphysics. In practice, I only used this to determine whether the cooling solve failed for any of the cells, in which case I made the simulation abort. There might be a better way to do this for Microphysics.

Reply:
Microphysics will throw an error if the substeps taken to integrate the network in a cell are more than some specific value. That error should result in quokka simulation being aborted, right?

Reply:
So then we wouldn't need these lines.

Reply:
If so, no. But it may not throw an error when running on GPU. Can you check the source code and/or with Mike about this?

Reply:
Following the conversation with Mike on Slack, how do we want to handle this for GPUs? For CPUs, the code should abort when it encounters `amrex:Error`, right?

Reply:
I added an abort loop which will work only on GPUs. How does it look

Reply:
That looks fine. Can you delete the `nsubstepsMF` variable?